Support Prost

nrc opened the pull request:
~~This can't be merged yet because it relies on path dependencies (waiting on changes to land in Prost)~~. Based on https://github.com/pingcap/rust-prometheus/pull/226

nrc commented:
This is now ready for review. PTAL @overvenus

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/pingcap/rust-prometheus?pullRequest=227) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/pingcap/rust-prometheus?pullRequest=227) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/pingcap/rust-prometheus?pullRequest=227) it.</sub>

overvenus commented:
Hi @nrc There are few warnings, would you mind address them? https://travis-ci.org/pingcap/rust-prometheus/jobs/518986584

nrc commented:
Hi addressed the warnings and rebased, but in the course of rebasing it became a bit more tricky to support both Prost and rust-protobuf. Since that is just a 'nice to have' for the migration, I've removed rust-protobuf support. I also removed support for having no codec, since I'm not sure why you would ever build without Prost (there is no version issue and protoc is not required)

nrc commented on src/encoder/pb.rs:
```diff
@@ -38,10 +38,12 @@ impl ProtobufEncoder {
 
 impl Encoder for ProtobufEncoder {
     fn encode<W: Write>(&self, metric_families: &[MetricFamily], writer: &mut W) -> Result<()> {
+        let mut buf = Vec::new();
```
For the sake of performance, we should change `encode` to take a `Bytes` buffer rather than a `Write` object, but I'm not sure how big a breaking change that would be.

nrc replied:
@overvenus what do you think?

nrc replied:
ping @overvenus  do you have thoughts on this?

breeswish replied:
I think we can leave it since ProtoBuf format will be removed in future (Prometheus has already dropped support for it).

nrc replied:
:+1:

breeswish commented on src/vec.rs:
```diff
@@ -48,14 +48,14 @@ impl<T: MetricVecBuilder> MetricVecCore<T> {
         m.set_name(self.desc.fq_name.clone());
         m.set_help(self.desc.help.clone());
-        m.set_field_type(self.metric_type);
+        m.set_field_type_(self.metric_type);
```
Why is there a `_` suffix?

nrc replied:
For enums, the `set_field_type` method is already generated by Prost, so we need to use a different name for the version we generate.

breeswish commented:
I'm fine with this PR. However I would like to see 0.6 carrying https://github.com/pingcap/rust-prometheus/pull/235 https://github.com/pingcap/rust-prometheus/pull/237 and https://github.com/pingcap/rust-prometheus/pull/238. Maybe leave this PR to 0.7 is better.

nrc commented:
@breeswish is it OK to merge this PR now?

breeswish commented:
Looks good. Could you fix the conflict?

breeswish commented on Cargo.toml:
```diff
@@ -17,31 +18,34 @@ travis-ci = { repository = "pingcap/rust-prometheus" }
 
 [dependencies]
-protobuf = { version = "2.0", optional = true }
+protobuf = "2.0"
```
Can you keep it optional? See https://github.com/pingcap/rust-prometheus/pull/216

nrc commented:
Hmm, after reading #216, it seems we don't need Prost support in Prometheus, so I think I can just close this PR.
